# Wrangle Pisa 2012
#### By Gabriela Sikora


### Abstract



### Table of Contents

<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
    <ul>
    <li><a href="#gathering">Gathering Data</a></li>
    <li><a href="#assessing1">Initial Assessment</a></li>
        <ul>
          <li><a href="#issues1">Identified Issues</a></li>
        </ul>
    <li><a href="#cleaning1">Cleaning Data</a></li>
    <li><a href="#assessing2">Final Assessment</a></li>
    </ul>
<li><a href="#sav">Store</a></li>
<li><a href="#conclusion">Conclusion</a></li>
<li><a href="#references">References</a></li>    
</ul>

<a id='intro'></a>
## Introduction



To begin, we first need to import all the relevant libraries:
#### Import

In [1]:
# import all packages
import numpy as np
import pandas as pd

<a id='wrangling'></a>
## Data Wrangling

In this section of the report, we will load in the data, check it for cleanliness, and then trim and clean the datasets for analysis. 

<a id='gathering'></a>
### Gathering Data

Unfortunately the Pisa dataset is too large to be uploaded to github, but if you would like to work with it yourself then place follow this link: https://s3.amazonaws.com/udacity-hosted-downloads/ud507/pisa2012.csv.zip 

In [2]:
# Encoding help found here: https://stackoverflow.com/questions/30462807/encoding-error-in-panda-read-csv

# Load in the Pisa dataframe
pisa = pd.read_csv('pisa2012.csv', encoding="cp1252", sep=',')

# Load in the Pisa dictionary
pisa_dict = pd.read_csv('pisadict2012.csv', encoding="cp1252", sep=',')

/Users/gabriela/Desktop/Code/myenv/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3044: DtypeWarning: Columns (15,16,17,21,22,23,24,25,26,30,31,36,37,45,65,123,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,284,285,286,287,288,289,290,291,292,293,294,295,296,297,298,299,300,301,302,303,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,391,392,393,394,395,396,397,398,399,400,401,402,403,475) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


<a id='assessing1'></a>
### First Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [3]:
pisa

Unnamed: 0      CNT  SUBNATIO  STRATUM      OECD         NC  SCHOOLID  \
0                1  Albania     80000  ALB0006  Non-OECD    Albania         1   
1                2  Albania     80000  ALB0006  Non-OECD    Albania         1   
2                3  Albania     80000  ALB0006  Non-OECD    Albania         1   
3                4  Albania     80000  ALB0006  Non-OECD    Albania         1   
4                5  Albania     80000  ALB0006  Non-OECD    Albania         1   
5                6  Albania     80000  ALB0006  Non-OECD    Albania         1   
6                7  Albania     80000  ALB0006  Non-OECD    Albania         1   
7                8  Albania     80000  ALB0006  Non-OECD    Albania         1   
8                9  Albania     80000  ALB0006  Non-OECD    Albania         1   
9               10  Albania     80000  ALB0005  Non-OECD    Albania         2   
10              11  Albania     80000  ALB0005  Non-OECD    Albania         2   
11              12  Albania     80000  ALB0005  Non-OECD    Albania         2   
12              13  Albania     80000  ALB0005  Non-OECD    Albania         2   
13              14  Albania     80000  ALB0005  Non-OECD    Albania         2   
14              15  Albania     80000  ALB0005  Non-OECD    Albania         2   
15              16  Albania     80000  ALB0005  Non-OECD    Albania         2   
16              17  Albania     80000  ALB0005  Non-OECD    Albania         2   
17              18  Albania     80000  ALB0005  Non-OECD    Albania         2   
18              19  Albania     80000  ALB0005  Non-OECD    Albania         2   
19              20  Albania     80000  ALB0005  Non-OECD    Albania         2   
20              21  Albania     80000  ALB0005  Non-OECD    Albania         2   
21              22  Albania     80000  ALB0005  Non-OECD    Albania         2   
22              23  Albania     80000  ALB0005  Non-OECD    Albania         2   
23              24  Albania     80000  ALB0005  Non-OECD    Albania         2   
24              25  Albania     80000  ALB0005  Non-OECD    Albania         2   
25              26  Albania     80000  ALB0005  Non-OECD    Albania         2   
26              27  Albania     80000  ALB0005  Non-OECD    Albania         2   
27              28  Albania     80000  ALB0005  Non-OECD    Albania         2   
28              29  Albania     80000  ALB0005  Non-OECD    Albania         2   
29              30  Albania     80000  ALB0005  Non-OECD    Albania         2   
...            ...      ...       ...      ...       ...        ...       ...   
485460      485461  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485461      485462  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485462      485463  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485463      485464  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485464      485465  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485465      485466  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485466      485467  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485467      485468  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485468      485469  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485469      485470  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485470      485471  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485471      485472  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485472      485473  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485473      485474  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485474      485475  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485475      485476  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485476      485477  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485477      485478  Vietnam   7040000  

In [4]:
pisa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Columns: 636 entries, Unnamed: 0 to VER_STU
dtypes: float64(250), int64(18), object(368)
memory usage: 2.3+ GB


In [5]:
pisa_dict

Unnamed: 0                                                  x
0           CNT                           Country code 3-character
1      SUBNATIO  Adjudicated sub-region code 7-digit code (3-di...
2       STRATUM  Stratum ID 7-character (cnt + region ID + orig...
3          OECD                                       OECD country
4            NC                       National Centre 6-digit Code
5      SCHOOLID  School ID 7-digit (region ID + stratum ID + 3-...
6       STIDSTD                                         Student ID
7       ST01Q01                                International Grade
8       ST02Q01                           National Study Programme
9       ST03Q01                                      Birth - Month
10      ST03Q02                                        Birth -Year
11      ST04Q01                                             Gender
12      ST05Q01                                   Attend <ISCED 0>
13      ST06Q01                                   Age at <ISCED 1>
14      ST07Q01                                 Repeat - <ISCED 1>
15      ST07Q02                                 Repeat - <ISCED 2>
16      ST07Q03                                 Repeat - <ISCED 3>
17      ST08Q01                          Truancy - Late for School
18      ST09Q01                    Truancy - Skip whole school day
19     ST115Q01           Truancy - Skip classes within school day
20      ST11Q01                                   At Home - Mother
21      ST11Q02                                   At Home - Father
22      ST11Q03                                 At Home - Brothers
23      ST11Q04                                  At Home - Sisters
24      ST11Q05                             At Home - Grandparents
25      ST11Q06                                   At Home - Others
26      ST13Q01                          Mother<Highest Schooling>
27      ST14Q01            Mother Qualifications - <ISCED level 6>
28      ST14Q02           Mother Qualifications - <ISCED level 5A>
29      ST14Q03           Mother Qualifications - <ISCED level 5B>
..          ...                                                ...
605    W_FSTR55           FINAL STUDENT REPLICATE BRR-FAY WEIGHT55
606    W_FSTR56           FINAL STUDENT REPLICATE BRR-FAY WEIGHT56
607    W_FSTR57           FINAL STUDENT REPLICATE BRR-FAY WEIGHT57
608    W_FSTR58           FINAL STUDENT REPLICATE BRR-FAY WEIGHT58
609    W_FSTR59           FINAL STUDENT REPLICATE BRR-FAY WEIGHT59
610    W_FSTR60           FINAL STUDENT REPLICATE BRR-FAY WEIGHT60
611    W_FSTR61           FINAL STUDENT REPLICATE BRR-FAY WEIGHT61
612    W_FSTR62           FINAL STUDENT REPLICATE BRR-FAY WEIGHT62
613    W_FSTR63           FINAL STUDENT REPLICATE BRR-FAY WEIGHT63
614    W_FSTR64           FINAL STUDENT REPLICATE BRR-FAY WEIGHT64
615    W_FSTR65           FINAL STUDENT REPLICATE BRR-FAY WEIGHT65
616    W_FSTR66           FINAL STUDENT REPLICATE BRR-FAY WEIGHT66
617    W_FSTR67           FINAL STUDENT REPLICATE BRR-FAY WEIGHT67
618    W_FSTR68           FINAL STUDENT REPLICATE BRR-FAY WEIGHT68
619    W_FSTR69           FINAL STUDENT REPLICATE BRR-FAY WEIGHT69
620    W_FSTR70           FINAL STUDENT REPLICATE BRR-FAY WEIGHT70
621    W_FSTR71           FINAL STUDENT REPLICATE BRR-FAY WEIGHT71
622    W_FSTR72           FINAL STUDENT REPLICATE BRR-FAY WEIGHT72
623    W_FSTR73           FINAL STUDENT REPLICATE BRR-FAY WEIGHT73
624    W_FSTR74           FINAL STUDENT REPLICATE BRR-FAY WEIGHT74
625    W_FSTR75           FINAL STUDENT REPLICATE BRR-FAY WEIGHT75
626    W_FSTR76           FINAL STUDENT REPLICATE BRR-FAY WEIGHT76
627    W_FSTR77           FINAL STUDENT REPLICATE BRR-FAY WEIGHT77
628    W_FSTR78           FINAL STUDENT REPLICATE BRR-FAY WEIGHT78
629    W_FSTR79           FINAL STUDENT REPLICATE BRR-FAY WEIGHT79
630    W_FSTR80           FINAL STUDENT REPLICATE BRR-FAY WEIGHT80
631    WVARSTRR           RANDOMIZED FINAL VARIANCE STRATUM (1-80)
632    VAR_UNIT                    RANDOMLY ASSIGNED

In [6]:
# Check for null values in relevant columns
columns = ['STIDSTD', 'ST01Q01', 'ST04Q01', 'ST13Q01', 
            'ST17Q01', 'ST28Q01', 'ST57Q01', 'ST57Q02', 'ST57Q03', 
            'ST57Q04', 'ST57Q05', 'MMINS', 'LMINS', 'SMINS',
            'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',  
            'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
            'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']

for x in columns:
    print(x, pisa[x].isna().sum())

STIDSTD 0
ST01Q01 0
ST04Q01 0
ST13Q01 27511
ST17Q01 42229
ST28Q01 11725
ST57Q01 184123
ST57Q02 215682
ST57Q03 201677
ST57Q04 205833
ST57Q05 195988
MMINS 202187
LMINS 202624
SMINS 214576
PV1MATH 0
PV2MATH 0
PV3MATH 0
PV4MATH 0
PV5MATH 0
PV1READ 0
PV2READ 0
PV3READ 0
PV4READ 0
PV5READ 0
PV1SCIE 0
PV2SCIE 0
PV3SCIE 0
PV4SCIE 0
PV5SCIE 0


In [7]:
pisa['CNT'].value_counts()

Mexico                      33806
Italy                       31073
Spain                       25313
Canada                      21544
Brazil                      19204
Australia                   14481
United Kingdom              12659
United Arab Emirates        11500
Switzerland                 11229
Qatar                       10966
Colombia                     9073
Finland                      8829
Belgium                      8597
Denmark                      7481
Jordan                       7038
Chile                        6856
Thailand                     6606
Japan                        6351
Chinese Taipei               6046
Peru                         6035
Slovenia                     5911
Argentina                    5908
Kazakhstan                   5808
Portugal                     5722
Indonesia                    5622
Singapore                    5546
Macao-China                  5335
Czech Republic               5327
Uruguay                      5315
Bulgaria      

In [8]:
pisa.duplicated().sum()

0

<a id='issues1'></a>
### Identified Issues
Here we can see some of the issues noticed in the dataframes. These quality issues will be cleaned in the following section.

1. Narrow down the dataframe by only looking at variables of interest since there are more variables than needed
2. Make the variable names readable since they are codes that can only be read with the dictionary
3. Create an average of all Math, Reading, and Science scores since all 5 scores for each are not necessary
4. Missing values in relevant columns ('Mother Highest Schooling', 'Father Highest Schooling', 'How many books at home', 'Out-of-School Study Time - Homework', 'Out-of-School Study Time - Guided Homework', 'Out-of-School Study Time - Personal Tutor','Out-of-School Study Time - Commercial Company', 'Out-of-School Study Time - With Parent', 'Learning time (minutes per week)- Mathematics', 'Learning time (minutes per week)  - test language', 'Learning time (minutes per week) - Science')
5. Create a total for Out-of-School Study Time and Learning time (minutes per week)
6. Single countries have various entries in the country column (Florida (USA), Massachusetts (USA), Connecticut (USA) and United States of America)


<a id='cleaning1'></a>
### First Cleaning

In [9]:
# Create a copy to preserve the original dataframe
pisa_clean = pisa.copy()

### Narrow down the dataframe by only looking at variables of interest

There are 636 columns available in the original dataframe, which is far more than what is necessary. We will rather look at just the following variables.

In [10]:
pisa_clean = pisa_clean[['STIDSTD', 'ST04Q01', 'ST13Q01', 'ST14Q01', 'ST14Q02', 'ST14Q03', 'ST14Q04', 
                         'ST17Q01', 'ST18Q01', 'ST18Q02', 'ST18Q03', 'ST18Q04', 
                         'ST28Q01', 
                         'ST57Q01', 'ST57Q02', 'ST57Q03', 'ST57Q04', 'ST57Q05', 'MMINS', 
                         'LMINS', 'SMINS', 'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 
                         'PV5MATH', 'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
                         'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']]       

In [11]:
# Check to see if the columns are correctly reduced
list(pisa_clean)

['STIDSTD',
 'ST04Q01',
 'ST13Q01',
 'ST14Q01',
 'ST14Q02',
 'ST14Q03',
 'ST14Q04',
 'ST17Q01',
 'ST18Q01',
 'ST18Q02',
 'ST18Q03',
 'ST18Q04',
 'ST28Q01',
 'ST57Q01',
 'ST57Q02',
 'ST57Q03',
 'ST57Q04',
 'ST57Q05',
 'MMINS',
 'LMINS',
 'SMINS',
 'PV1MATH',
 'PV2MATH',
 'PV3MATH',
 'PV4MATH',
 'PV5MATH',
 'PV1READ',
 'PV2READ',
 'PV3READ',
 'PV4READ',
 'PV5READ',
 'PV1SCIE',
 'PV2SCIE',
 'PV3SCIE',
 'PV4SCIE',
 'PV5SCIE']

### Create an average of all Math, Reading, and Science scores

For each of the Math, Reading, and Science scores, there are 5 scores provided. The variation between these scores is not needed, so rather we will look at the average for each of these major categories and remove the individual scores.

In [12]:
# Create average of Math scores
pisa_clean['Average Math Score']    = (pisa_clean['PV1MATH'] + pisa_clean['PV2MATH'] + pisa_clean['PV3MATH'] + pisa_clean['PV4MATH'] + pisa_clean['PV5MATH']) / 5

# Create average of Reading scores
pisa_clean['Average Reading Score'] = (pisa_clean['PV1READ'] + pisa_clean['PV2READ'] + pisa_clean['PV3READ'] + pisa_clean['PV4READ'] + pisa_clean['PV5READ']) / 5

# Create average of Science scores
pisa_clean['Average Science Score'] = (pisa_clean['PV1SCIE'] + pisa_clean['PV2SCIE'] + pisa_clean['PV3SCIE'] + pisa_clean['PV4SCIE'] + pisa_clean['PV5SCIE']) / 5

# Create average score
pisa_clean['Average Total Score']   = (pisa_clean['Average Math Score'] + pisa_clean['Average Reading Score'] + pisa_clean['Average Science Score'] +pisa_clean['Average Math Score']) / 3

In [13]:
# Remove the individual scores for Math, Reading, and Science
pisa_clean.drop(columns=['PV1MATH',  'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',   
                         'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ',  'PV5READ',
                         'PV1SCIE',  'PV2SCIE', 'PV3SCIE',  'PV4SCIE', 'PV5SCIE'], inplace=True)

In [14]:
# Test to see accuracy of the Math score
print(pisa_clean['Average Math Score'].describe())
print('-'*22)

# Test to see accuracy of the Reading score
print(pisa_clean['Average Reading Score'].describe())
print('-'*22)

# Test to see accuracy of the Science score
print(pisa_clean['Average Science Score'].describe())

# Test to see accuracy of the Total score
print(pisa_clean['Average Total Score'].describe())

count    485490.000000
mean        469.651234
std         100.786610
min          54.767080
25%         396.019620
50%         465.734520
75%         540.123060
max         903.107960
Name: Average Math Score, dtype: float64
----------------------
count    485490.000000
mean        472.006964
std          98.863310
min           6.445400
25%         405.044200
50%         475.477980
75%         542.831195
max         849.359740
Name: Average Reading Score, dtype: float64
----------------------
count    485490.000000
mean        475.808094
std          97.998470
min          25.158540
25%         405.762800
50%         475.512860
75%         546.381920
max         857.832900
Name: Average Science Score, dtype: float64
count    485490.000000
mean        629.039175
std         128.839689
min         105.496473
25%         536.312167
50%         627.318087
75%         721.099990
max        1123.882273
Name: Average Total Score, dtype: float64


In [15]:
pisa_clean.sample(10)

STIDSTD ST04Q01                                ST13Q01 ST14Q01  \
437159     4154    Male                       <ISCED level 2>       No   
436644     3639    Male                      <ISCED level 3A>       No   
19543      3301  Female                       <ISCED level 1>      NaN   
369007      412  Female                      <ISCED level 3A>       No   
371359     2764  Female                  <ISCED level 3B, 3C>       No   
423315      540  Female                      <ISCED level 3A>       No   
180950     5276    Male                  <ISCED level 3B, 3C>      NaN   
71769     16504    Male  She did not complete <ISCED level 1>      NaN   
434157     1152  Female                      <ISCED level 3A>       No   
15284     10542    Male                      <ISCED level 3A>       No   

       ST14Q02 ST14Q03 ST14Q04                               ST17Q01 ST18Q01  \
437159      No      No      No                      <ISCED level 2>       No   
436644      No      No     NaN                     <ISCED level 3A>       No   
19543      NaN     NaN     NaN                                   NaN     NaN   
369007     Yes      No      No                     <ISCED level 3A>       No   
371359      No      No      No                 <ISCED level 3B, 3C>       No   
423315      No      No     Yes                     <ISCED level 3A>       No   
180950     Yes     NaN     NaN                     <ISCED level 3A>       No   
71769      NaN     NaN     NaN  He did not complete <ISCED level 1>      NaN   
434157      No      No     Yes                     <ISCED level 3A>       No   
15284       No      No      No                     <ISCED level 3A>      NaN   

       ST18Q02  ... ST57Q03 ST57Q04 ST57Q05  MMINS  LMINS  SMINS  \
437159      No  ...     0.0     1.0     0.0  135.0  135.0   45.0   
436644      No  ...     NaN     NaN     NaN    NaN    NaN    NaN   
19543      NaN  ...     NaN     NaN     NaN    NaN    NaN    NaN   
369007     Yes  ...     NaN     NaN     NaN    NaN    NaN    NaN   
371359      No  ...     0.0     0.0     1.0  180.0  180.0  180.0   
423315      No  ...     NaN     NaN     NaN    NaN    NaN    NaN   
180950     Yes  ...     NaN     NaN     NaN    NaN    NaN    NaN   
71769      NaN  ...     NaN     NaN     NaN    NaN    NaN    NaN   
434157     Yes  ...     NaN     NaN     NaN    NaN    NaN    NaN   
15284      NaN  ...     NaN     NaN     NaN    NaN    NaN    NaN   

        Average Math Score  Average Reading Score  Average Science Score  \
437159           355.59276              326.93474              365.47930   
436644           514.65184              463.34562              499.29132   
19543            234.15640              247.19326              219.73144   
369007           609.37060              771.67624              722.99512   
371359           452.80420              531.07970              535.84484   
423315           604.22964              564.52016              600.46622   
180950           561.07648              524.29344              566.98992   
71769            293.97882              282.26640              317.26966   
434157           492.29630              509.63334              477.75084   
15284            352.71068              364.22520              343.47258   

        Average Total Score  
437159           467.866520  
436644           663.980207  
19543            311.745833  
369007           904.470853  
371359           657.510980  
423315           791.148553  
180950           737.812107  
71769            395.831233  
434157           657.325593  
15284            471.039713  

[10 rows x 25 columns]

### Make the variable names readable

In the original dataframe, the variable names are all codes and unreadable without the dictionary. Therefore, we will change them to accurately reflect the information within the column.

In [16]:
# Rename columns
pisa_clean.rename({'STIDSTD': 'Student ID',
                   'ST04Q01': 'Gender', 
                   'ST13Q01': 'Mother - Highest Schooling', 
                   'ST14Q01': 'Mother Qualifications - ISCED level 6',
                   'ST14Q02': 'Mother Qualifications - ISCED level 5A',
                   'ST14Q03': 'Mother Qualifications - ISCED level 5B',
                   'ST14Q04': 'Mother Qualifications - ISCED level 4',
                   'ST17Q01': 'Father - Highest Schooling',
                   'ST18Q01': 'Father Qualifications - ISCED level 6',
                   'ST18Q02': 'Father Qualifications - ISCED level 5A',
                   'ST18Q03': 'Father Qualifications - ISCED level 5B',
                   'ST18Q04': 'Father Qualifications - ISCED level 4',
                   'ST28Q01': 'How many books at home', 
                   'ST57Q01': 'Out-of-School Study Time - Homework',
                   'ST57Q02': 'Out-of-School Study Time - Guided Homework',
                   'ST57Q03': 'Out-of-School Study Time - Personal Tutor',
                   'ST57Q04': 'Out-of-School Study Time - Commercial Company',
                   'ST57Q05': 'Out-of-School Study Time - With Parent',
                   'MMINS': 'Learning time (minutes per week)- <Mathematics>',
                   'LMINS': 'Learning time (minutes per week)  - <test language>',
                   'SMINS': 'Learning time (minutes per week) - <Science>'}, axis='columns', inplace=True)

In [17]:
# Check to see if the columns are correctly renamed
list(pisa_clean)

['Student ID',
 'Gender',
 'Mother - Highest Schooling',
 'Mother Qualifications - ISCED level 6',
 'Mother Qualifications - ISCED level 5A',
 'Mother Qualifications - ISCED level 5B',
 'Mother Qualifications - ISCED level 4',
 'Father - Highest Schooling',
 'Father Qualifications - ISCED level 6',
 'Father Qualifications - ISCED level 5A',
 'Father Qualifications - ISCED level 5B',
 'Father Qualifications - ISCED level 4',
 'How many books at home',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning time (minutes per week)- <Mathematics>',
 'Learning time (minutes per week)  - <test language>',
 'Learning time (minutes per week) - <Science>',
 'Average Math Score',
 'Average Reading Score',
 'Average Science Score',
 'Average Total Score']

# Null

In [18]:
pisa_clean.replace(["NaN"], np.nan, inplace = True)
pisa_clean.dropna(inplace = True)

### Education
Multiple columns with the single variable. 

http://uis.unesco.org/sites/default/files/documents/international-standard-classification-of-education-isced-2011-en.pdf

In [19]:
father_edu = pisa_clean.copy()

In [20]:
father_edu[['<ISCED level 1>', 
            '<ISCED level 2>', 
            '<ISCED level 3A>', 
            '<ISCED level 3B, 3C>', 
            'He did not complete <ISCED level 1>']] = pd.get_dummies(father_edu['Father - Highest Schooling'])

In [21]:
father_edu.rename({'He did not complete <ISCED level 1>': '<ISCED level 0>',
                   '<ISCED level 3A>': '<ISCED level 3>', 
                   '<ISCED level 3B, 3C>': '<ISCED level 3>',
                   'Father Qualifications - ISCED level 4': '<ISCED level 4>',
                   'Father Qualifications - ISCED level 5A': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 5B': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 6': '<ISCED level 6>'}, axis='columns', inplace=True)

In [22]:
father_edu['<ISCED level 0>'] = father_edu['<ISCED level 0>'].astype(str)
father_edu['<ISCED level 1>'] = father_edu['<ISCED level 1>'].astype(str)
father_edu['<ISCED level 2>'] = father_edu['<ISCED level 2>'].astype(str)
father_edu['<ISCED level 3>'] = father_edu['<ISCED level 3>'].astype(str)

In [23]:
father_edu['<ISCED level 0>'] = father_edu['<ISCED level 0>'].replace(regex='1', value='<ISCED level 0>')
father_edu['<ISCED level 1>'] = father_edu['<ISCED level 1>'].replace(regex='1', value='<ISCED level 1>')
father_edu['<ISCED level 2>'] = father_edu['<ISCED level 2>'].replace(regex='1', value='<ISCED level 2>')
father_edu['<ISCED level 3>'] = father_edu['<ISCED level 3>'].replace(regex='1', value='<ISCED level 3>')
father_edu['<ISCED level 4>'] = father_edu['<ISCED level 4>'].replace(regex='Yes', value='<ISCED level 4>')
father_edu['<ISCED level 5>'] = father_edu['<ISCED level 5>'].replace(regex='Yes', value='<ISCED level 5>')
father_edu['<ISCED level 6>'] = father_edu['<ISCED level 6>'].replace(regex='Yes', value='<ISCED level 6>')

In [24]:
father_edu.sample(10)

Student ID  Gender             Mother - Highest Schooling  \
148573        2991    Male  She did not complete <ISCED level 1>    
23211         1061  Female                      <ISCED level 3A>    
62421         7156    Male                      <ISCED level 3A>    
183432        7758  Female                      <ISCED level 3A>    
153207        7625  Female                       <ISCED level 2>    
48751         7365    Male                  <ISCED level 3B, 3C>    
249940        9351  Female                      <ISCED level 3A>    
433166         161  Female                  <ISCED level 3B, 3C>    
235390        3364  Female                      <ISCED level 3A>    
262258       21669  Female                       <ISCED level 2>    

       Mother Qualifications - ISCED level 6  \
148573                                    No   
23211                                     No   
62421                                     No   
183432                                    No   
153207                                    No   
48751                                     No   
249940                                    No   
433166                                    No   
235390                                    No   
262258                                    No   

       Mother Qualifications - ISCED level 5A  \
148573                                     No   
23211                                      No   
62421                                      No   
183432                                    Yes   
153207                                     No   
48751                                      No   
249940                                    Yes   
433166                                     No   
235390                                     No   
262258                                     No   

       Mother Qualifications - ISCED level 5B  \
148573                                     No   
23211                                      No   
62421                                      No   
183432                                     No   
153207                                     No   
48751                                     Yes   
249940                                     No   
433166                                     No   
235390                                     No   
262258                                     No   

       Mother Qualifications - ISCED level 4 Father - Highest Schooling  \
148573                                    No           <ISCED level 1>    
23211                                     No           <ISCED level 2>    
62421                                     No          <ISCED level 3A>    
183432                                    No      <ISCED level 3B, 3C>    
153207                                    No           <ISCED level 2>    
48751                                     No          <ISCED level 3A>    
249940                                    No          <ISCED level 3A>    
433166                                    No      <ISCED level 3B, 3C>    
235390                                   Yes          <ISCED level 3A>    
262258                                   Yes           <ISCED level 2>    

       <ISCED level 6>  <ISCED level 5>  ...  \
148573              No               No  ...   
23211               No               No  ...   
62421               No               No  ...   
183432              No               No  ...   
153207              No               No  ...   
48751               No  <ISCED level 5>  ...   
249940              No  <ISCED level 5>  ...   
433166              No               No  ...   
235390              No               No  ...   
262258              No               No  ...   

       Learning time (minutes per week) - <Science> Average Math Score  \
148573                                        275.0          453.58312   
23211                                         180.0          370.62626   
62421                                          90.0          2

In [25]:
father_edu = pd.melt(father_edu, 
                     id_vars=['Student ID', 'Gender',
                            'Mother - Highest Schooling', 
                            'Mother Qualifications - ISCED level 6', 
                            'Mother Qualifications - ISCED level 5A',
                            'Mother Qualifications - ISCED level 5B',
                            'Mother Qualifications - ISCED level 4',
                            'How many books at home',
                            'Out-of-School Study Time - Homework',
                            'Out-of-School Study Time - Guided Homework',
                            'Out-of-School Study Time - Personal Tutor',
                            'Out-of-School Study Time - Commercial Company',
                            'Out-of-School Study Time - With Parent',
                            'Learning time (minutes per week)- <Mathematics>',
                            'Learning time (minutes per week)  - <test language>',
                            'Learning time (minutes per week) - <Science>',
                            'Average Math Score',
                            'Average Reading Score',
                            'Average Science Score',
                            'Average Total Score'], var_name='Completed_edu - Father', value_name='Education - Father')


In [26]:
father_edu.sample(10)

Student ID  Gender Mother - Highest Schooling  \
102897         1624    Male           <ISCED level 2>    
739764          844  Female           <ISCED level 2>    
529019        19085    Male          <ISCED level 3A>    
208301         3808    Male      <ISCED level 3B, 3C>    
181960         5608  Female      <ISCED level 3B, 3C>    
1127220         612  Female          <ISCED level 3A>    
1083130        3984    Male           <ISCED level 2>    
1038628          58  Female           <ISCED level 1>    
911452         3640    Male           <ISCED level 2>    
288177         1367  Female          <ISCED level 3A>    

        Mother Qualifications - ISCED level 6  \
102897                                     No   
739764                                     No   
529019                                     No   
208301                                     No   
181960                                     No   
1127220                                    No   
1083130                                    No   
1038628                                    No   
911452                                     No   
288177                                     No   

        Mother Qualifications - ISCED level 5A  \
102897                                      No   
739764                                      No   
529019                                     Yes   
208301                                      No   
181960                                      No   
1127220                                     No   
1083130                                     No   
1038628                                     No   
911452                                      No   
288177                                     Yes   

        Mother Qualifications - ISCED level 5B  \
102897                                      No   
739764                                      No   
529019                                      No   
208301                                      No   
181960                                      No   
1127220                                     No   
1083130                                     No   
1038628                                     No   
911452                                      No   
288177                                      No   

        Mother Qualifications - ISCED level 4 How many books at home  \
102897                                    Yes          26-100 books    
739764                                     No          26-100 books    
529019                                     No          26-100 books    
208301                                     No           11-25 books    
181960                                    Yes         201-500 books    
1127220                                   Yes          26-100 books    
1083130                                    No         201-500 books    
1038628                                    No            0-10 books    
911452                                     No          26-100 books    
288177                                     No    More than 500 books   

         Out-of-School Study Time - Homework  \
102897                                  10.0   
739764                                   1.0   
529019                                   2.0   
208301                                   0.0   
181960                                   5.0   
1127220                                  8.0   
1083130                                  8.0   
1038628                                  5.0   
911452                                   1.0   
288177                                  20.0   

         Out-of-School Study Time - Guided Homework  ...  \
102897                                          0.0  ...   
739764                                          1.0  ...   
529019                                          0.0  ...   
208301                                          0.0  ...   
181960                                          3.0  ...   
1127220                                       

In [27]:
father_edu = father_edu[(father_edu['Education - Father'] != '0')] 
father_edu = father_edu[(father_edu['Education - Father'] != 'No')]

In [28]:
father_edu.sample(70)

Student ID  Gender             Mother - Highest Schooling  \
898853        1437    Male                      <ISCED level 3A>    
833448        3464  Female                      <ISCED level 3A>    
99656         4593  Female                      <ISCED level 3A>    
77407        30829    Male                       <ISCED level 2>    
993125         627  Female                  <ISCED level 3B, 3C>    
17527         3474  Female                      <ISCED level 3A>    
808729        6942    Male                      <ISCED level 3A>    
537072         408  Female                      <ISCED level 3A>    
104444        1026    Male                      <ISCED level 3A>    
328036         453  Female                  <ISCED level 3B, 3C>    
85069         3662    Male                  <ISCED level 3B, 3C>    
107345        4216    Male                      <ISCED level 3A>    
439888         797    Male                       <ISCED level 2>    
99373         3970    Male                       <ISCED level 2>    
378986        7285    Male                  <ISCED level 3B, 3C>    
992658        3846    Male                  <ISCED level 3B, 3C>    
108399        1372    Male                      <ISCED level 3A>    
919563        1598  Female                  <ISCED level 3B, 3C>    
1977          5139  Female                      <ISCED level 3A>    
700088       18215    Male                      <ISCED level 3A>    
129579       15316    Male                      <ISCED level 3A>    
513730         604    Male                      <ISCED level 3A>    
31689         2461    Male                      <ISCED level 3A>    
428822         521    Male                      <ISCED level 3A>    
32418         4155    Male                      <ISCED level 3A>    
238343        5590    Male                      <ISCED level 3A>    
861377         296  Female                      <ISCED level 3A>    
446107        4939  Female                  <ISCED level 3B, 3C>    
35920         4548    Male                       <ISCED level 2>    
82207         3928  Female                  <ISCED level 3B, 3C>    
...            ...     ...                                    ...   
911044        2562  Female                      <ISCED level 3A>    
89316         3563  Female                      <ISCED level 3A>    
446980        3124  Female                  <ISCED level 3B, 3C>    
79874         1490  Female                      <ISCED level 3A>    
40             235    Male                  <ISCED level 3B, 3C>    
946270         471    Male                  <ISCED level 3B, 3C>    
195039         252    Male                      <ISCED level 3A>    
81083          382    Male                      <ISCED level 3A>    
86913          684  Female                       <ISCED level 2>    
103396        3046  Female                       <ISCED level 1>    
71803        15814    Male                      <ISCED level 3A>    
921157        1661  Female                  <ISCED level 3B, 3C>    
13861         6436    Male  She did not complete <ISCED level 1>    
26299         6590    Male                      <ISCED level 3A>    
563691        5055    Male                      <ISCED level 3A>    
24460         1839    Male                      <ISCED level 3A>    
790984        3437  Female                      <ISCED level 3A>    
845512        1024    Male                  <ISCED level 3B, 3C>    
51709         3991    Male                      <ISCED level 3A>    
784965        1730  Female                      <ISCED level 3A>    
472699        1162  Female                      <ISCED level 3A>    
404105        1562  Female                       <ISCED level 2>    
20132        10937    Male                      <ISCED level 3A>    
51370         3033  Female                      <ISCED level 3A>    
55159         2957  Female                      <ISCED level 3A>    
359638        3864    Male                      <ISCED level 3A>    
848995

In [29]:
father_edu = father_edu[(father_edu['Completed_edu - Father'] != 'Father - Highest Schooling')] 

In [30]:
father_edu['Education - Father'].value_counts()

<ISCED level 3>    84016
<ISCED level 5>    48776
<ISCED level 4>    26281
<ISCED level 2>    21813
<ISCED level 6>     6672
<ISCED level 1>     5901
<ISCED level 0>     2264
Name: Education - Father, dtype: int64

In [31]:
father_edu.drop(columns=['Completed_edu - Father'], axis=1, inplace=True)

In [32]:
father_edu.sample(10)

Student ID  Gender Mother - Highest Schooling  \
448249        1688  Female          <ISCED level 3A>    
915143       11434    Male          <ISCED level 3A>    
263884        4499  Female          <ISCED level 3A>    
898380         173  Female          <ISCED level 3A>    
845628        1349  Female          <ISCED level 3A>    
819685       15710    Male          <ISCED level 3A>    
268369       10274  Female          <ISCED level 3A>    
784806        1282    Male           <ISCED level 2>    
886402         656  Female      <ISCED level 3B, 3C>    
721831        2425    Male          <ISCED level 3A>    

       Mother Qualifications - ISCED level 6  \
448249                                    No   
915143                                    No   
263884                                    No   
898380                                    No   
845628                                   Yes   
819685                                    No   
268369                                    No   
784806                                    No   
886402                                    No   
721831                                    No   

       Mother Qualifications - ISCED level 5A  \
448249                                    Yes   
915143                                    Yes   
263884                                    Yes   
898380                                     No   
845628                                    Yes   
819685                                     No   
268369                                     No   
784806                                     No   
886402                                     No   
721831                                     No   

       Mother Qualifications - ISCED level 5B  \
448249                                     No   
915143                                     No   
263884                                    Yes   
898380                                     No   
845628                                     No   
819685                                     No   
268369                                     No   
784806                                     No   
886402                                     No   
721831                                     No   

       Mother Qualifications - ISCED level 4 How many books at home  \
448249                                    No         101-200 books    
915143                                    No           11-25 books    
263884                                   Yes          26-100 books    
898380                                   Yes          26-100 books    
845628                                    No         201-500 books    
819685                                    No    More than 500 books   
268369                                    No         101-200 books    
784806                                    No            0-10 books    
886402                                   Yes          26-100 books    
721831                                    No            0-10 books    

        Out-of-School Study Time - Homework  \
448249                                  5.0   
915143                                  1.0   
263884                                  6.0   
898380                                  1.0   
845628                                  1.0   
819685                                 12.0   
268369                                  5.0   
784806                                  2.0   
886402                                  4.0   
721831                                  3.0   

        Out-of-School Study Time - Guided Homework  ...  \
448249                                         0.0  ...   
915143                                         0.0  ...   
263884                                         2.0  ...   
898380                                         1.0  ...   
845628                                         1.0  ...   
819685                                        10.0  ...   
268369                                         1.0  ...   

In [33]:
father_edu.duplicated().sum()

6901

In [34]:
# https://stackoverflow.com/questions/33042777/removing-duplicates-from-pandas-dataframe-with-condition-for-retaining-original
father_edu['Education - Father'] = father_edu['Education - Father'].astype('category')

father_edu['Education - Father'] = father_edu['Education - Father'].cat.set_categories(['<ISCED level 0>', 
                                                                                        '<ISCED level 1>', 
                                                                                        '<ISCED level 2>', 
                                                                                        '<ISCED level 3>',
                                                                                        '<ISCED level 4>', 
                                                                                        '<ISCED level 5>', 
                                                                                        '<ISCED level 6>'], ordered=True)
father_edu.sort_values(['Education - Father'], inplace=True)
father_edu_clean = father_edu.drop_duplicates(subset='Student ID', keep='last')

In [35]:
father_edu_clean.duplicated().sum()

0

In [36]:
father_edu.shape

(195723, 21)

### Mother Edu

In [37]:
mother_edu = father_edu_clean.copy()

In [38]:
mother_edu.head()

Student ID  Gender             Mother - Highest Schooling  \
1055573        6301  Female  She did not complete <ISCED level 1>    
1048110       16914    Male                      <ISCED level 3A>    
1047956       16487    Male                      <ISCED level 3A>    
1047644       15611  Female                      <ISCED level 3A>    
1047154       14175    Male                      <ISCED level 3A>    

        Mother Qualifications - ISCED level 6  \
1055573                                    No   
1048110                                    No   
1047956                                    No   
1047644                                    No   
1047154                                    No   

        Mother Qualifications - ISCED level 5A  \
1055573                                     No   
1048110                                    Yes   
1047956                                     No   
1047644                                     No   
1047154                                    Yes   

        Mother Qualifications - ISCED level 5B  \
1055573                                     No   
1048110                                     No   
1047956                                     No   
1047644                                     No   
1047154                                     No   

        Mother Qualifications - ISCED level 4 How many books at home  \
1055573                                    No            0-10 books    
1048110                                    No          26-100 books    
1047956                                    No           11-25 books    
1047644                                    No          26-100 books    
1047154                                    No         201-500 books    

         Out-of-School Study Time - Homework  \
1055573                                  0.0   
1048110                                  0.0   
1047956                                  1.0   
1047644                                  2.0   
1047154                                  1.0   

         Out-of-School Study Time - Guided Homework  ...  \
1055573                                         0.0  ...   
1048110                                         0.0  ...   
1047956                                         1.0  ...   
1047644                                         0.0  ...   
1047154                                         1.0  ...   

         Out-of-School Study Time - Commercial Company  \
1055573                                            0.0   
1048110                                            0.0   
1047956                                            0.0   
1047644                                            0.0   
1047154                                            0.0   

         Out-of-School Study Time - With Parent  \
1055573                                     1.0   
1048110                                     0.0   
1047956                                     1.0   
1047644                                     0.0   
1047154                                     1.0   

         Learning time (minutes per week)- <Mathematics>  \
1055573                                            540.0   
1048110                                            280.0   
1047956                                            400.0   
1047644                                            375.0   
1047154                                            400.0   

         Learning time (minutes per week)  - <test language>  \
1055573                                              720.0     
1048110                                              280.0     
1047956                                              400.0     
1047644                                              375.0     
1047154                                                0.0     

         Learning time (minutes per week) - <Science>  Average Math Score  \
1055573                                          90.0           301.69028   
1048110                                         280.

In [39]:
mother_edu['Mother - Highest Schooling'].value_counts()

<ISCED level 3A>                         15433
<ISCED level 3B, 3C>                      3798
<ISCED level 2>                           3605
<ISCED level 1>                            532
She did not complete <ISCED level 1>       152
Name: Mother - Highest Schooling, dtype: int64

In [40]:
pisa_clean['Mother - Highest Schooling'].value_counts()

<ISCED level 3A>                         63289
<ISCED level 3B, 3C>                     23470
<ISCED level 2>                          19733
<ISCED level 1>                           5149
She did not complete <ISCED level 1>      2353
Name: Mother - Highest Schooling, dtype: int64

In [41]:
mother_edu[['<ISCED level 1>', 
            '<ISCED level 2>', 
            '<ISCED level 3A>', 
            '<ISCED level 3B, 3C>', 
            'She did not complete <ISCED level 1>']] = pd.get_dummies(mother_edu['Mother - Highest Schooling'])

In [42]:
mother_edu.rename({'She did not complete <ISCED level 1>': '<ISCED level 0>',
                   '<ISCED level 3A>': '<ISCED level 3>', 
                   '<ISCED level 3B, 3C>': '<ISCED level 3>',
                   'Mother Qualifications - ISCED level 4': '<ISCED level 4>',
                   'Mother Qualifications - ISCED level 5A': '<ISCED level 5>',
                   'Mother Qualifications - ISCED level 5B': '<ISCED level 5>',
                   'Mother Qualifications - ISCED level 6': '<ISCED level 6>'}, axis='columns', inplace=True)

In [43]:
mother_edu['<ISCED level 0>'] = mother_edu['<ISCED level 0>'].astype(str)
mother_edu['<ISCED level 1>'] = mother_edu['<ISCED level 1>'].astype(str)
mother_edu['<ISCED level 2>'] = mother_edu['<ISCED level 2>'].astype(str)
mother_edu['<ISCED level 3>'] = mother_edu['<ISCED level 3>'].astype(str)

In [44]:
mother_edu['<ISCED level 0>'] = mother_edu['<ISCED level 0>'].replace(regex='1', value='<ISCED level 0>')
mother_edu['<ISCED level 1>'] = mother_edu['<ISCED level 1>'].replace(regex='1', value='<ISCED level 1>')
mother_edu['<ISCED level 2>'] = mother_edu['<ISCED level 2>'].replace(regex='1', value='<ISCED level 2>')
mother_edu['<ISCED level 3>'] = mother_edu['<ISCED level 3>'].replace(regex='1', value='<ISCED level 3>')
mother_edu['<ISCED level 4>'] = mother_edu['<ISCED level 4>'].replace(regex='Yes', value='<ISCED level 4>')
mother_edu['<ISCED level 5>'] = mother_edu['<ISCED level 5>'].replace(regex='Yes', value='<ISCED level 5>')
mother_edu['<ISCED level 6>'] = mother_edu['<ISCED level 6>'].replace(regex='Yes', value='<ISCED level 6>')

In [45]:
mother_edu.sample(10)

Student ID  Gender Mother - Highest Schooling  <ISCED level 6>  \
870885       18748    Male          <ISCED level 3A>                No   
210761        4342  Female          <ISCED level 3A>   <ISCED level 6>   
248339       11624    Male          <ISCED level 3A>   <ISCED level 6>   
302277       22418    Male          <ISCED level 3A>                No   
178084        2244  Female          <ISCED level 3A>   <ISCED level 6>   
281689        9405    Male      <ISCED level 3B, 3C>                No   
874111       27372    Male      <ISCED level 3B, 3C>                No   
385728       20642    Male          <ISCED level 3A>                No   
363127       13981  Female          <ISCED level 3A>                No   
217681        3824  Female          <ISCED level 3A>   <ISCED level 6>   

        <ISCED level 5>  <ISCED level 5>  <ISCED level 4>  \
870885               No               No               No   
210761  <ISCED level 5>  <ISCED level 5>               No   
248339  <ISCED level 5>               No               No   
302277  <ISCED level 5>               No               No   
178084  <ISCED level 5>               No               No   
281689               No  <ISCED level 5>  <ISCED level 4>   
874111  <ISCED level 5>               No               No   
385728               No               No  <ISCED level 4>   
363127  <ISCED level 5>               No               No   
217681  <ISCED level 5>  <ISCED level 5>  <ISCED level 4>   

       How many books at home  Out-of-School Study Time - Homework  \
870885           11-25 books                                   9.0   
210761           11-25 books                                  10.0   
248339            0-10 books                                   3.0   
302277    More than 500 books                                 30.0   
178084         101-200 books                                   4.0   
281689          26-100 books                                   1.0   
874111          26-100 books                                   4.0   
385728         201-500 books                                   8.0   
363127         201-500 books                                   6.0   
217681         201-500 books                                  20.0   

        Out-of-School Study Time - Guided Homework  ...  Average Math Score  \
870885                                         0.0  ...           569.02166   
210761                                         0.0  ...           661.16996   
248339                                         1.0  ...           425.85294   
302277                                         0.0  ...           509.90028   
178084                                         2.0  ...           524.07696   
281689                                         0.0  ...           569.87850   
874111                                         1.0  ...           456.85466   
385728                                         1.0  ...           438.16016   
363127                                         0.0  ...           586.85932   
217681                                         0.0  ...           595.19398   

        Average Reading Score  Average Science Score  Average Total Score  \
870885              544.18166              600.18650           760.803827   
210761              634.18112              605.68816           854.069733   
248339              381.46704              399.04886           544.073927   
302277              542.81838              544.79670           702.471880   
178084              504.62918              497.70608           683.496393   
281689              557.01280              558.13130           751.633700   
874111              452.84012              426.65052           597.733320   
385728              413.54480              435.78890           575.218007   
363127              574.13128              639.44424           795.764720   
217681              654.35664              552.62962           799.124740   

        Education - Father  <ISC

In [46]:
mother_edu = pd.melt(mother_edu, 
                     id_vars=['Student ID', 'Gender',
                            'How many books at home',
                            'Out-of-School Study Time - Homework',
                            'Out-of-School Study Time - Guided Homework',
                            'Out-of-School Study Time - Personal Tutor',
                            'Out-of-School Study Time - Commercial Company',
                            'Out-of-School Study Time - With Parent',
                            'Learning time (minutes per week)- <Mathematics>',
                            'Learning time (minutes per week)  - <test language>',
                            'Learning time (minutes per week) - <Science>',
                            'Average Math Score',
                            'Average Reading Score',
                            'Average Science Score',
                            'Average Total Score',
                             'Education - Father'], var_name='Completed_edu - Mother', value_name='Education - Mother')


In [47]:
mother_edu.sample(10)

Student ID  Gender How many books at home  \
231297       11342  Female           11-25 books    
129125       10887  Female         101-200 books    
181962        4682    Male          26-100 books    
95094        24426  Female            0-10 books    
216823       17743    Male          26-100 books    
12461         2578    Male          26-100 books    
175667        7029    Male         101-200 books    
1955         14789    Male          26-100 books    
195689       20872    Male         201-500 books    
172777        9845    Male         101-200 books    

        Out-of-School Study Time - Homework  \
231297                                  3.0   
129125                                 12.0   
181962                                  3.0   
95094                                   6.0   
216823                                  2.0   
12461                                   2.0   
175667                                  1.0   
1955                                   10.0   
195689                                 15.0   
172777                                  2.0   

        Out-of-School Study Time - Guided Homework  \
231297                                         2.0   
129125                                         0.0   
181962                                         0.0   
95094                                          0.0   
216823                                         0.0   
12461                                          0.0   
175667                                         0.0   
1955                                           9.0   
195689                                         0.0   
172777                                         0.0   

        Out-of-School Study Time - Personal Tutor  \
231297                                        2.0   
129125                                        0.0   
181962                                        0.0   
95094                                         0.0   
216823                                        0.0   
12461                                         0.0   
175667                                        0.0   
1955                                          0.0   
195689                                        0.0   
172777                                        0.0   

        Out-of-School Study Time - Commercial Company  \
231297                                            1.0   
129125                                            0.0   
181962                                            0.0   
95094                                             0.0   
216823                                            0.0   
12461                                             0.0   
175667                                            0.0   
1955                                              0.0   
195689                                            1.0   
172777                                            0.0   

        Out-of-School Study Time - With Parent  \
231297                                     1.0   
129125                                     1.0   
181962                                     0.0   
95094                                      0.0   
216823                                     0.0   
12461                                      0.0   
175667                                     0.0   
1955                                       0.0   
195689                                     1.0   
172777                                     1.0   

        Learning time (minutes per week)- <Mathematics>  \
231297                                            225.0   
129125                                            300.0   
181962                                            200.0   
95094                                             150.0   
216823                                            300.0   
12461                                             270.0   
175667                                            360.0   
1955                                              240.0   
195689           

In [48]:
mother_edu = mother_edu[(mother_edu['Education - Mother'] != '0')] 
mother_edu = mother_edu[(mother_edu['Education - Mother'] != 'No')]

In [49]:
mother_edu.sample(70)

Student ID  Gender How many books at home  \
172195       20134    Male           11-25 books    
9113          8694    Male         101-200 books    
45665         4651  Female         101-200 books    
4844         16070    Male           11-25 books    
111685        5071    Male          26-100 books    
14880         5783  Female           11-25 books    
8258         19102  Female         101-200 books    
861          14332  Female          26-100 books    
194508       31064  Female         201-500 books    
91758        20337    Male          26-100 books    
58281        15362    Male         201-500 books    
104568       20813    Male            0-10 books    
176148       10444    Male         101-200 books    
180561        4949  Female         101-200 books    
60176         7485  Female            0-10 books    
211594        1517    Male         101-200 books    
8782         18018    Male           11-25 books    
78152        18847    Male           11-25 books    
185538        1977  Female           11-25 books    
1637         28889    Male          26-100 books    
97504        24047  Female            0-10 books    
101376       11288  Female          26-100 books    
70314         3451    Male          26-100 books    
19997         1165  Female         101-200 books    
57503        20559    Male          26-100 books    
62087         4689    Male    More than 500 books   
102204        5922    Male         101-200 books    
90695         1981  Female          26-100 books    
23299         1628  Female           11-25 books    
185484        3875  Female    More than 500 books   
...            ...     ...                    ...   
10114        16585    Male          26-100 books    
64802         7033    Male         101-200 books    
180592        6177    Male         201-500 books    
20928         9799    Male            0-10 books    
17465          312  Female    More than 500 books   
209701       12061    Male         101-200 books    
84315        17451    Male           11-25 books    
174135       16462    Male    More than 500 books   
172180       21168    Male          26-100 books    
21238         3841  Female    More than 500 books   
179661        7333    Male         101-200 books    
19599         9704    Male         101-200 books    
3257         17558    Male          26-100 books    
24102        24091  Female    More than 500 books   
167001       16420  Female            0-10 books    
142246       25468  Female          26-100 books    
134480       14305    Male    More than 500 books   
176622        8804    Male           11-25 books    
23017         3045  Female    More than 500 books   
15091           27  Female         101-200 books    
144493       24770  Female          26-100 books    
6960         13997  Female          26-100 books    
11858        19595  Female         101-200 books    
21972        10263  Female          26-100 books    
51423         9172  Female         101-200 books    
165826       24883  Female         201-500 books    
8246         16925    Male          26-100 books    
22998          187    Male          26-100 books    
15269         4559  Female         101-200 books    
21947         8525    Male            0-10 books    

        Out-of-School Study Time - Homework  \
172195                                  4.0   
9113                                    3.0   
45665                                  17.0   
4844                                    2.0   
111685                                  5.0   
14880                                   2.0   
8258                                    4.0   
861                                    10.0   
194508                                 20.0   
91758                                   1.0   
58281                                   6.0   
104568                                 10.0   
176148                                  5.0   
180561                                  5.0   
60176           

In [50]:
mother_edu = mother_edu[(mother_edu['Completed_edu - Mother'] != 'Mother - Highest Schooling')] 

In [51]:
mother_edu['Education - Mother'].value_counts()

<ISCED level 3>    19231
<ISCED level 5>    14533
<ISCED level 4>     5094
<ISCED level 2>     3605
<ISCED level 6>     2611
<ISCED level 1>      532
<ISCED level 0>      152
Name: Education - Mother, dtype: int64

In [52]:
mother_edu.drop(columns=['Completed_edu - Mother'], axis=1, inplace=True)

In [53]:
mother_edu.sample(10)

Student ID  Gender How many books at home  \
172201       20555  Female         101-200 books    
196447       18619    Male         101-200 books    
68874         2823  Female           11-25 books    
110542        8527  Female          26-100 books    
101135       20265    Male          26-100 books    
171354       30893    Male         201-500 books    
168533       13055  Female    More than 500 books   
172258       19745    Male           11-25 books    
55825        17802    Male         201-500 books    
93311        15497    Male         201-500 books    

        Out-of-School Study Time - Homework  \
172201                                 15.0   
196447                                 10.0   
68874                                   5.0   
110542                                 10.0   
101135                                  1.0   
171354                                  2.0   
168533                                  5.0   
172258                                  2.0   
55825                                   0.0   
93311                                   7.0   

        Out-of-School Study Time - Guided Homework  \
172201                                         2.0   
196447                                         0.0   
68874                                          5.0   
110542                                         5.0   
101135                                         0.0   
171354                                         0.0   
168533                                         2.0   
172258                                         2.0   
55825                                          0.0   
93311                                          2.0   

        Out-of-School Study Time - Personal Tutor  \
172201                                        0.0   
196447                                        7.0   
68874                                         3.0   
110542                                        7.0   
101135                                        0.0   
171354                                        0.0   
168533                                        0.0   
172258                                        0.0   
55825                                         1.0   
93311                                         0.0   

        Out-of-School Study Time - Commercial Company  \
172201                                            0.0   
196447                                            0.0   
68874                                             2.0   
110542                                            2.0   
101135                                            0.0   
171354                                            0.0   
168533                                            0.0   
172258                                            0.0   
55825                                             0.0   
93311                                             0.0   

        Out-of-School Study Time - With Parent  \
172201                                     0.0   
196447                                     0.0   
68874                                      2.0   
110542                                    10.0   
101135                                     0.0   
171354                                     0.0   
168533                                     2.0   
172258                                     0.0   
55825                                      0.0   
93311                                      2.0   

        Learning time (minutes per week)- <Mathematics>  \
172201                                            375.0   
196447                                            240.0   
68874                                             265.0   
110542                                            200.0   
101135                                            250.0   
171354                                            100.0   
168533                                            150.0   
172258                                            195.0   
55825            

In [75]:
mother_edu['Student ID'].duplicated().sum()

22238

In [55]:
# mother_edu.replace(["NaN"], np.nan, inplace = True)
# mother_edu.dropna(inplace = True)

In [77]:
mother_edu['Education - Mother'] = mother_edu['Education - Mother'].astype('category')

mother_edu['Education - Mother'] = mother_edu['Education - Mother'].cat.set_categories(['<ISCED level 0>', 
                                                                                        '<ISCED level 1>', 
                                                                                        '<ISCED level 2>', 
                                                                                        '<ISCED level 3>',
                                                                                        '<ISCED level 4>', 
                                                                                        '<ISCED level 5>', 
                                                                                        '<ISCED level 6>'], ordered=True)
mother_edu.sort_values(['Education - Mother'], inplace=True)
mother_edu_clean = mother_edu.drop_duplicates(subset='Student ID', keep='last')

In [78]:
mother_edu['Student ID'].duplicated().sum()

22238

In [58]:
mother_edu.shape

(45758, 17)

In [59]:
mother_edu_clean.shape

(23520, 17)

In [60]:
mother_edu_clean.duplicated().sum()

0

In [61]:
mother_edu_clean.sample(10)

Student ID  Gender How many books at home  \
65070        12115    Male            0-10 books    
61797        24192  Female            0-10 books    
79846        10937    Male         101-200 books    
45279         6206  Female         201-500 books    
104101       24966  Female         101-200 books    
46814         2350    Male          26-100 books    
82751         9239  Female         101-200 books    
166578       14448  Female          26-100 books    
100805       30765    Male          26-100 books    
63463         4231    Male    More than 500 books   

        Out-of-School Study Time - Homework  \
65070                                   1.0   
61797                                   2.0   
79846                                   9.0   
45279                                   1.0   
104101                                  4.0   
46814                                  12.0   
82751                                   4.0   
166578                                 19.0   
100805                                  7.0   
63463                                   2.0   

        Out-of-School Study Time - Guided Homework  \
65070                                          1.0   
61797                                          0.0   
79846                                          0.0   
45279                                          1.0   
104101                                         0.0   
46814                                          8.0   
82751                                          2.0   
166578                                        19.0   
100805                                         5.0   
63463                                          0.0   

        Out-of-School Study Time - Personal Tutor  \
65070                                         0.0   
61797                                         0.0   
79846                                         0.0   
45279                                         2.0   
104101                                        0.0   
46814                                         4.0   
82751                                         0.0   
166578                                        0.0   
100805                                        2.0   
63463                                         0.0   

        Out-of-School Study Time - Commercial Company  \
65070                                             0.0   
61797                                             0.0   
79846                                             0.0   
45279                                             2.0   
104101                                            0.0   
46814                                             6.0   
82751                                             0.0   
166578                                            0.0   
100805                                            0.0   
63463                                             0.0   

        Out-of-School Study Time - With Parent  \
65070                                      0.0   
61797                                      0.0   
79846                                      0.0   
45279                                      0.0   
104101                                     0.0   
46814                                     12.0   
82751                                      1.0   
166578                                     2.0   
100805                                     5.0   
63463                                      0.0   

        Learning time (minutes per week)- <Mathematics>  \
65070                                             120.0   
61797                                             240.0   
79846                                             300.0   
45279                                             250.0   
104101                                            165.0   
46814                                             240.0   
82751                                             300.0   
166578                                            165.0   
100805           

In [62]:
pisa_clean = mother_edu_clean.copy()

In [63]:
pisa_clean.shape

(23520, 17)

In [64]:
pisa_clean = mother_edu.copy()

In [65]:
pisa_clean.shape

(45758, 17)

### 4. Missing values in relevant columns 
('Mother Highest Schooling', 'Father Highest Schooling', 'How many books at home', 'Out-of-School Study Time - Homework', 'Out-of-School Study Time - Guided Homework', 'Out-of-School Study Time - Personal Tutor','Out-of-School Study Time - Commercial Company', 'Out-of-School Study Time - With Parent', 'Learning time (minutes per week)- Mathematics', 'Learning time (minutes per week)  - test language', 'Learning time (minutes per week) - Science')

Due to the amount of rows in this dataframe, we can comfortably remove the rows with missing values.

In [66]:

# pisa_clean.replace(["NaN"], np.nan, inplace = True)
# pisa_clean.dropna(inplace = True)

In [67]:
# Check to see if null values correctly dropped
pisa_clean

Student ID  Gender How many books at home  \
235181        1647    Male            0-10 books    
212053       14280    Male           11-25 books    
212039       18805  Female            0-10 books    
212037       18913    Male           11-25 books    
212019       12822    Male           11-25 books    
212011        9143    Male           11-25 books    
212009       10479    Male            0-10 books    
211995       16142  Female           11-25 books    
211994       16295  Female            0-10 books    
211993       17846    Male            0-10 books    
211987       17761    Male            0-10 books    
211974       14327  Female            0-10 books    
211972       13635    Male            0-10 books    
211957       15437  Female         101-200 books    
211952       15337    Male            0-10 books    
211838       21333    Male          26-100 books    
212089       19354  Female            0-10 books    
211817        6640    Male            0-10 books    
212140       22540  Female           11-25 books    
212206       17718  Female            0-10 books    
213611       27839    Male          26-100 books    
213067       22804  Female           11-25 books    
212698       24651  Female          26-100 books    
212471       12825    Male            0-10 books    
212468       13375    Male          26-100 books    
212465       13203  Female           11-25 books    
212448       13900    Male           11-25 books    
212369       10904  Female            0-10 books    
212367       11498    Male          26-100 books    
212294       10648  Female         101-200 books    
...            ...     ...                    ...   
42790        14415    Male          26-100 books    
42792        14413  Female          26-100 books    
42793        14401    Male          26-100 books    
42795        14395  Female            0-10 books    
42796        14277  Female          26-100 books    
42800        14165    Male            0-10 books    
42804        14999  Female         101-200 books    
42805        14957    Male         101-200 books    
42777        15714    Male          26-100 books    
42776        15718  Female         101-200 books    
42775        15785  Female    More than 500 books   
42773        15055  Female          26-100 books    
42739        12665    Male          26-100 books    
42740        12624  Female            0-10 books    
42742        12566    Male          26-100 books    
42744        12504    Male           11-25 books    
42745        12496  Female    More than 500 books   
42749        20493    Male         101-200 books    
42750        14156  Female          26-100 books    
42809        14580    Male         101-200 books    
42752        15082    Male          26-100 books    
42756        15237    Male          26-100 books    
42762        15456  Female          26-100 books    
42764        15537    Male         201-500 books    
42766        15544    Male         101-200 books    
42767        16106    Male         101-200 books    
42769        15992  Female          26-100 books    
42771        15940    Male           11-25 books    
42754        15089  Female    More than 500 books   
23582        23884    Male    More than 500 books   

        Out-of-School Study Time - Homework  \
235181                                  2.0   
212053                                  3.0   
212039                                  0.0   
212037                                  2.0   
212019                                 30.0   
212011                                  4.0   
212009                                  2.0   
211995                                  5.0   
211994                                  1.0   
211993                                  2.0   
211987                                  1.0   
211974                                  1.0   
211972                                  2.0   
211957                                  8.0   
211952          

In [68]:
# Confirm with smaller shape of dataframe
pisa_clean.shape

(45758, 17)

### 5. Create a total for Out-of-School Study Time and Learning time (minutes per week)



In [69]:
# Create total of Out-of-School Study Time
pisa_clean['Out-of-School Study Time - Total'] = pisa_clean['Out-of-School Study Time - Homework'] + pisa_clean['Out-of-School Study Time - Guided Homework'] + pisa_clean['Out-of-School Study Time - Personal Tutor'] + pisa_clean['Out-of-School Study Time - Commercial Company'] + pisa_clean['Out-of-School Study Time - With Parent']

# Create total of Learning time (minutes per week)
pisa_clean['Learning time (minutes per week) - Total'] = pisa_clean['Learning time (minutes per week)- <Mathematics>'] + pisa_clean['Learning time (minutes per week)  - <test language>'] + pisa_clean['Learning time (minutes per week) - <Science>'] 
                                                                                                                 

In [70]:
list(pisa_clean)

['Student ID',
 'Gender',
 'How many books at home',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning time (minutes per week)- <Mathematics>',
 'Learning time (minutes per week)  - <test language>',
 'Learning time (minutes per week) - <Science>',
 'Average Math Score',
 'Average Reading Score',
 'Average Science Score',
 'Average Total Score',
 'Education - Father',
 'Education - Mother',
 'Out-of-School Study Time - Total',
 'Learning time (minutes per week) - Total']

<a id='assessing2'></a>
### Final Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [71]:
pisa_clean

Student ID  Gender How many books at home  \
235181        1647    Male            0-10 books    
212053       14280    Male           11-25 books    
212039       18805  Female            0-10 books    
212037       18913    Male           11-25 books    
212019       12822    Male           11-25 books    
212011        9143    Male           11-25 books    
212009       10479    Male            0-10 books    
211995       16142  Female           11-25 books    
211994       16295  Female            0-10 books    
211993       17846    Male            0-10 books    
211987       17761    Male            0-10 books    
211974       14327  Female            0-10 books    
211972       13635    Male            0-10 books    
211957       15437  Female         101-200 books    
211952       15337    Male            0-10 books    
211838       21333    Male          26-100 books    
212089       19354  Female            0-10 books    
211817        6640    Male            0-10 books    
212140       22540  Female           11-25 books    
212206       17718  Female            0-10 books    
213611       27839    Male          26-100 books    
213067       22804  Female           11-25 books    
212698       24651  Female          26-100 books    
212471       12825    Male            0-10 books    
212468       13375    Male          26-100 books    
212465       13203  Female           11-25 books    
212448       13900    Male           11-25 books    
212369       10904  Female            0-10 books    
212367       11498    Male          26-100 books    
212294       10648  Female         101-200 books    
...            ...     ...                    ...   
42790        14415    Male          26-100 books    
42792        14413  Female          26-100 books    
42793        14401    Male          26-100 books    
42795        14395  Female            0-10 books    
42796        14277  Female          26-100 books    
42800        14165    Male            0-10 books    
42804        14999  Female         101-200 books    
42805        14957    Male         101-200 books    
42777        15714    Male          26-100 books    
42776        15718  Female         101-200 books    
42775        15785  Female    More than 500 books   
42773        15055  Female          26-100 books    
42739        12665    Male          26-100 books    
42740        12624  Female            0-10 books    
42742        12566    Male          26-100 books    
42744        12504    Male           11-25 books    
42745        12496  Female    More than 500 books   
42749        20493    Male         101-200 books    
42750        14156  Female          26-100 books    
42809        14580    Male         101-200 books    
42752        15082    Male          26-100 books    
42756        15237    Male          26-100 books    
42762        15456  Female          26-100 books    
42764        15537    Male         201-500 books    
42766        15544    Male         101-200 books    
42767        16106    Male         101-200 books    
42769        15992  Female          26-100 books    
42771        15940    Male           11-25 books    
42754        15089  Female    More than 500 books   
23582        23884    Male    More than 500 books   

        Out-of-School Study Time - Homework  \
235181                                  2.0   
212053                                  3.0   
212039                                  0.0   
212037                                  2.0   
212019                                 30.0   
212011                                  4.0   
212009                                  2.0   
211995                                  5.0   
211994                                  1.0   
211993                                  2.0   
211987                                  1.0   
211974                                  1.0   
211972                                  2.0   
211957                                  8.0   
211952          

In [72]:
pisa_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45758 entries, 235181 to 23582
Data columns (total 19 columns):
Student ID                                             45758 non-null int64
Gender                                                 45758 non-null object
How many books at home                                 45758 non-null object
Out-of-School Study Time - Homework                    45758 non-null float64
Out-of-School Study Time - Guided Homework             45758 non-null float64
Out-of-School Study Time - Personal Tutor              45758 non-null float64
Out-of-School Study Time - Commercial Company          45758 non-null float64
Out-of-School Study Time - With Parent                 45758 non-null float64
Learning time (minutes per week)- <Mathematics>        45758 non-null float64
Learning time (minutes per week)  - <test language>    45758 non-null float64
Learning time (minutes per week) - <Science>           45758 non-null float64
Average Math Score              

In [73]:
pisa_clean.shape

(45758, 19)

<a id='sav'></a>
## Store

In [74]:
# Store the clean DataFrame in a CSV file
pisa_clean.to_csv("pisa_df.csv", index=False, encoding='utf8')

<a id='conclusion'></a>
## Conclusion

<a id='references'></a>
## References